### ISSUE TITLE
Fire `request`-event on every request

### ISSUE THREAD

**Issue by hardnold** [hapijs/hapi#4335]
#### Support plan

* *is this issue currently blocking your project?* (yes/no):
* *is this issue affecting a production system?* (yes/no):

#### Context

* *node version*: 14.7.6
* *module version*:  All Versions >=18.0.0
* *environment* (e.g. node, browser, native): node
* *used with* (e.g. hapi application, another framework, standalone, ...): standalone
* *any other relevant information*: -

#### What problem are you trying to solve?

I was wondering why the `request`-event only fires in error-cases. [Before 18.0.0 the `request`-event always had the `error`-tag set](https://github.com/hapijs/hapi/issues/3898), which apparently resulted in the `request`-event being triggered at every request (which actually sounds like the correct behaviour to me). [This comment in the same issue](https://github.com/hapijs/hapi/issues/3898#issuecomment-454236132) also states that the request-event is supposed to be only fired in error-cases (which indeed is true since 18.0.0 and can be reproduced fairly easy with a failing request). 

This behaviour, however, doesn't make sense to me at all due to the name of the event being `request`. By looking at the data the event-handler receives (e.g. the `channel`), it would be easy to determine whether the request resulted in an error or not and also leads to the assumption that this is the reason for the data being present.

[The documentation](https://hapi.dev/api?v=20.2.1#-request-event) also is not very clear about this behaviour, as it just states:

> The 'request' event type emits internal request events generated by the framework as well as application events logged with [request.log()](https://hapi.dev/api?v=20.2.1#request.log()).

There is no hint about this event only being triggered in error-cases.

I'd suggest to either
- fire the event on every request (which would be my preferred scenario due to the name and the data of the event) or
- add some clear information about this behaviour to the documentation.

I spent the whole day yesterday to investigate why a custom plugin, which reacts on the request-event, wasn't working anymore as expected, some clear documentation would have probably reduced the time spent to an hour tops or so ;)

**Comment by kanongil** [comment]
I agree that this is a bit confusing compared to the `response` event, which occurs at a fixed point.

The `request` event is not a single event, but a *channel*, and is for events that relate to an active request. This is mostly relevant for errors, but can also be used for app / plugin events in the `app` subchannel when you do `request.log()`. We could potentially add some non-error events, eg. a `'lifecycle', 'routed'` event that could occur right after the "Route lookup" lifecycle stage. Or `'lifecycle', 'incoming'` right after "onRequest".

The `response` event could have been a sub event on the internal `request` event channel. I expect that it is common enough to warrant a completely separate event name.

FYI, renaming to `error` won't work, as the event relates to an active `request` and then it could be confused with a server-level error and constrain app event to errors. Regardless of the name, this approach is a breaking change.

Maybe the best solution is to update the docs so that it is more clear that `request` is a channel, vs. `response` that happens once?

**Comment by hardnold** [comment]
Thanks for clarification, this is indeed a bit confusing ;)

In my particular case, I updated our plugin to react on the `response`-event now, since this suffices our needs (for logging request-specific information) and also comes with all the request-related data in the request-argument.

But still, adding more details to the docs would be very helpful in my opinion.

**Comment by kanongil** [comment]
@hardnold Glad this can make some sense. This is probably mostly an issue for legacy code that relies on the buggy behavior. Any chance you want to create a PR with a better explanation?
